### PR TITLE
Update cmod I/O

### LINKED ISSUE #1151: Upper limit on token lengths?
Currently working on a few changes that involve the tokenizer used to parse the various catalog files, one thing I've noticed is it currently allows unlimited token lengths, which would eventually exhaust memory. In order to tighten things up, I am planning to add an upper limit: if the token length is exceeded, the token will be reported as an error and parsing will bail out.

As an initial suggestion, I'm considering putting the limit to 1024 characters (which basically correspond to UTF-8 code units), perhaps with a lower limit (64) for tokens that aren't quote-delimited strings.

Any use cases you can see for supporting longer tokens?

### PR DESCRIPTION
**Pull request by ajtribick**
* Replace the tokenizer implementation in the celmodel code with the one used for other file types, move the tokenizer to celutil
* Use the binaryread.hpp/binarywrite.hpp routines for parsing binary CMOD files
* Enforce maximum token length limit of 1024 characters - resolves #1151 
* Disallow creation of surrogate pair characters (U+D800 to U+DFFF) via Unicode escape sequences
* Detect and reject invalid UTF-8 sequences in the tokenizer

### CODE REVIEW

File: src/celmodel/modelfile.cpp [hunk 1]
```diff
@@ -2127,251 +1635,332 @@ BinaryModelLoader::loadVertices(const Mesh::VertexDescription& vertexDesc,
...
-                out.write(cdata, 4);
+                result = out.write(cdata, 4).good();
                 break;
             default:
                 assert(0);
                 break;
             }
+
+            if (!result)
+            {
```
Comment (375gnu) [inline src/celmodel/modelfile.cpp:1812]: could you use the same style everywhere?

**Review by 375gnu — CHANGES_REQUESTED** [review]

**Comment by 375gnu** [comment]
did you think about replacing material::color with celutil::color? for me it's annoying to have 2 similar implementations

File: src/celmodel/modelfile.cpp (hunk 1 above)
Comment (ajtribick) [reply src/celmodel/modelfile.cpp:1812]: What style are you referring to here?

File: src/celmodel/modelfile.cpp [hunk 2]
```diff
@@ -8,535 +8,110 @@
...
-
-        } // Switch
-
-        if (!hasError() && !newToken.isValid())
-        {
-            m_nextChar = readChar();
-        }
-    }
+ public:
+    virtual ~ModelWriter() {};
```
[the remark this answers is not in the document]
Comment (ajtribick) [reply src/celmodel/modelfile.cpp:108]: Leftover from moving this class out of modelfile.h. Fixed

File: src/celutil/utf8.h [hunk 3]
```diff
@@ -136,4 +136,26 @@ class Greek
...
+{
+public:
+    UTF8Validator() = default;
+    ~UTF8Validator() = default;
+
+    bool check(char c);
+    bool check(unsigned char c);
+
+private:
+    enum class State {
```
[the remark this answers is not in the document]
Comment (ajtribick) [reply src/celutil/utf8.h:149]: Done.

**Comment by ajtribick** [comment]
Yes, that's definitely something I wanted to look at... I've been doing some further tidying-up of the model-related code which I'll raise in a future PR.

File: src/celmodel/modelfile.cpp (hunk 1 above)
Comment (375gnu) [reply src/celmodel/modelfile.cpp:1812]: there are many places where you do `if (...) { return ...; }`. Here you use multiple lines.

Comment (ajtribick) [reply src/celmodel/modelfile.cpp]: Got it. Fixed.

**Review by 375gnu — APPROVED** [review]

**Pull request merged**